hail/python/hailtop/aiogoogle/client/storage_client.py

```diff
@@ -60,6 +68,206 @@ async def _wait_closed(self):
+            self._size -= n
+        assert self._offset == new_offset
+
+    def chunks(self, chunk_size: int) -> Iterator[bytes]:
+        assert not self._iterating
+        self._iterating = True
+        remaining = chunk_size
+        i = 0
+        while remaining > 0:
+            b = self._buffers[i]
+            n = len(b)
+            if n <= remaining:
+                yield b
+                remaining -= n
+                i += 1
+            else:
+                yield b[:remaining]
+                break
+        self._iterating = False
+
+
+class _TaskManager:
+    def __init__(self, coro):
+        self._coro = coro
+        self._task = None
+
+    async def __aenter__(self) -> '_TaskManager':
+        self._task = asyncio.create_task(self._coro)
+        return self._task
+
+    async def __aexit__(self,
+                        exc_type: Optional[Type[BaseException]],
+                        exc_val: Optional[BaseException],
+                        exc_tb: Optional[TracebackType]) -> None:
+        if not self._task.done():
+            if exc_val:
+                self._task.cancel()
+                try:
+                    await self._task
+                except:
+                    _, exc, _ = sys.exc_info()
+                    if exc is not exc_val:
+                        log.warning('dropping preempted task exception', exc_info=True)
+            else:
+                await self._task
+
+
+class ResumableInsertObjectStream(WritableStream):
+    def __init__(self, session: Session, session_url: str, chunk_size: int):
+        super().__init__()
+        self._session = session
+        self._session_url = session_url
+        self._write_buffer = _WriteBuffer()
+        self._broken = False
+        self._done = False
+        self._chunk_size = chunk_size
+
+    @staticmethod
+    def _range_upper(range):
+        range = range.split('/', 1)[0]
+        split_range = range.split('-')
+        assert len(split_range) == 2
+        return int(split_range[1])
+
+    async def _write_chunk_1(self):
+        if self._closed:
+            total_size = self._write_buffer.offset() + self._write_buffer.size()
+            total_size_str = str(total_size)
+        else:
+            total_size = None
+            total_size_str = '*'
+
+        if self._broken:
+            # note: this retries
+            resp = self._session.put(self._session_url,
+                                     headers={
+                                         'Content-Length': '0',
+                                         'Content-Range': f'bytes */{total_size_str}'
+                                     },
+                                     raise_for_status=False)
+            if resp.status >= 200 or resp.status < 300:
+                assert self._closed and self._write_buffer.size() == 0
+                self._done = True
+                return
+            if resp.status == 308:
+                range = resp.header.get('Range')
```

Comment:
It might be helpful to point to docs for this:
https://cloud.google.com/storage/docs/json_api/v1/objects/insert
https://cloud.google.com/storage/docs/performing-resumable-uploads#status-check